### PR TITLE
auto_save_optimization

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Resolves: https://github.com/musescore/MuseScore/issues/13661

### CODE REVIEW

**Comment by cbjeukendrup** [comment]
MS3 had another optimisation: when there were no changes since the last autosave, a new autosave won't be performed. Of course, that does not reduce the annoyance that autosave might cause when working actively on a score, but it does save a lot of energy when the user leaves a score open while being away from their computer.

**Comment by MarcSabatella** [comment]
I had opened an issue a while back regarding the lack of that particular "don't autosave if no changes" optimization - https://github.com/musescore/MuseScore/issues/13661

File: src/project/internal/notationproject.cpp [hunk 1]
```diff
@@ -128,18 +128,36 @@ void NotationProject::setupProject()
     m_engravingProject = EngravingProject::create();
     m_engravingProject->setFileInfoProvider(std::make_shared<ProjectFileInfoProvider>(this));
 
+    auto listenNotationChanges = [this](INotationPtr notation) {
+        INotationWeakPtr weakPtr = notation;
+
+        notation->undoStack()->changesChannel().onReceive(this, [this, weakPtr](const ScoreChangesRange&) {
```
Comment (cbjeukendrup) [inline src/project/internal/notationproject.cpp:134]: Since there is essentially only one undo stack for the whole project (I mean, there is one _underlying_ UndoStack, namely in the `MasterScore`), it should be enough to listen only for changes from the MasterScore / MasterNotation. The notifications from the excerpts are only duplicates.

Comment (RomanPudashkin) [reply src/project/internal/notationproject.cpp]: Thanks! Actually, this code was only working because it was inside the loop. changesChannel() was not available for MasterNotation in setupProject() (because m_score had not yet been initialized), but became available after receiving some changes in the excerpt list. Now this code should be more robust

File: src/project/internal/notationproject.cpp [hunk 2]
```diff
@@ -128,18 +128,36 @@ void NotationProject::setupProject()
...
+            INotationPtr ptr = weakPtr.lock();
+            bool isStackClean = ptr && ptr->undoStack()->isStackClean();
+            setNeedSave(!isStackClean);
```
[the remark this answers is not in the document]
Comment (RomanPudashkin) [reply src/project/internal/notationproject.cpp]: Unfortunately, the rules here are more complicated. We need to mark the project as saved if the user clears the undo stack. But I fixed the problem you found, thanks!

File: src/project/internal/notationproject.cpp [hunk 3]
```diff
@@ -830,33 +873,39 @@ void NotationProject::markAsSaved(const io::path_t& path)
...
+void NotationProject::setNeedSave(bool needSave)
+{
+    mu::engraving::MasterScore* score = m_masterNotation->masterScore();
+    if (!score) {
+        return;
     }
 
-    m_needSaveNotificationBlocked = false;
-    m_needSaveNotification.notify();
+    setNeedAutoSave(needSave);
```
Comment (cbjeukendrup) [inline src/project/internal/notationproject.cpp:892]: Maybe this logic is still not perfect; what about the following situation?
1. User saves file
2. User performs action X -> marked as "need save" and "need autosave"
3. Wait until autosave does its work -> marked as "not need autosave"
4. User performs action Y -> marked as "need autosave"
5. User undoes action Y -> expected: marked as "not need autosave"; actual: still marked as "need autosave"

It's quite an edge case, so I'm not sure if we really need to worry about it.

File: src/project/internal/notationproject.cpp [hunk 4]
```diff
@@ -830,33 +873,39 @@ void NotationProject::markAsSaved(const io::path_t& path)
...
     m_isNewlyCreated = false;
 
-    m_needSaveNotificationBlocked = true;
+    setNeedSave(false);
+
+    setPath(path);
 
-    // Mark engraving project as saved
-    m_masterNotation->setSaved(true);
+    m_hasNonUndoStackChanges = false;
```
Comment (cbjeukendrup) [inline src/project/internal/notationproject.cpp:880]: Shouldn't this happen inside `setNeedSave(false)`?

File: src/project/internal/notationproject.cpp (hunk 3 above)
Comment (RomanPudashkin) [reply src/project/internal/notationproject.cpp:892]: This looks like a very edge case that would probably require a significant over-complication of the implementation (unless I'm missing something)

Comment (cbjeukendrup) [reply src/project/internal/notationproject.cpp:890]: There might be a semi-easy fix: inside Libmscore, each state of the UndoStack gets an ID. If you expose this ID via INotationUndoStack, you can use it in NotationProject: you just store the state ID of the last save and of the last autosave. Once you have that info, things are easy:
```c++
needSave == (m_lastSaveUndoStackStateId != undoStack()->currentStateId()) || m_hasNonUndoStackChangesSinceLastSave;
needAutosave == (m_lastAutosaveUndoStackStateId != undoStack()->currentStateId()) || m_hasNonUndoStackChangesSinceLastAutosave;
```
You only need to be careful with the value of `m_lastSaveUndoStackStateId` and `m_lastAutosaveUndoStackStateId` when the score has never been saved yet or when the undo stack is empty. 

Anyway, perhaps that's too much of a complication, given how little value it brings.

**Review by cbjeukendrup — APPROVED** [review]

File: src/engraving/libmscore/undo.h [hunk 5]
```diff
@@ -943,6 +943,8 @@ class RemoveExcerpt : public UndoCommand
     void undo(EditData*) override;
     void redo(EditData*) override;
 
+    std::vector<const EngravingObject*> objectItems() const override;
```
Comment (cbjeukendrup) [inline src/engraving/libmscore/undo.h:946]: What is the purpose of this change? And why is it needed only for RemoveExcerpt and not for AddExcerpt?

Comment (RomanPudashkin) [reply src/engraving/libmscore/undo.h:948]: objectItems() specifies which engraving objects will be changed when using this command. Without this change, we can't send data to changesChannel() (as ScoreChangesRange will be considered invalid), so the score will not be marked as unsaved. See this code in Score::endCmd:
    
   ```
 if (!rollback && range.isValid()) {
        changesChannel().send(range);
    }
```

I added the same change for AddExcerpt, thanks

**Comment by DmitryArefiev** [comment]
Tested #13661 on Win10, Mac13, LinuxUbuntu 22.04 - **FIXED**

File: src/engraving/libmscore/undo.h (hunk 5 above)
Comment (cbjeukendrup) [reply src/engraving/libmscore/undo.h:948]: Hm... I see. But isn't that a bit dangerous? It looks like there are more actions without object items, but simply every action should trigger the file to become unsaved.

Comment (RomanPudashkin) [reply src/engraving/libmscore/undo.h:948]: I moved the range.isValid() check directly to PlaybackModel. Now even if we failed to resolve ScoreChangesRange from some command, we still mark the project as unsaved

Comment (cbjeukendrup) [reply src/engraving/libmscore/undo.h:948]: Seems good to me!
